ta/pkcs11/src/sanitize_object.c

```diff
@@ -0,0 +1,466 @@
+}
+
+/* Counterpart of serialize_indirect_attribute() */
+static uint32_t sanitize_indirect_attr(struct obj_attrs **dst,
+				       struct pkcs11_attribute_head *cli_ref,
+				       char *cur)
+{
+	struct obj_attrs *obj2 = NULL;
+	enum pkcs11_rc rc = PKCS11_CKR_OK;
+	enum pkcs11_class_id class = get_class(*dst);
+
+	if (class == PKCS11_CKO_UNDEFINED_ID)
+		return PKCS11_CKR_GENERAL_ERROR;
+
+	/*
+	 * Serialized attributes: current applicable only the key templates
+	 * which are tables of attributes.
+	 */
+	switch (cli_ref->id) {
+	case PKCS11_CKA_WRAP_TEMPLATE:
+	case PKCS11_CKA_UNWRAP_TEMPLATE:
+	case PKCS11_CKA_DERIVE_TEMPLATE:
+		break;
+	default:
+		return PKCS11_RV_NOT_FOUND;
+	}
+	/* Such attributes are expected only for keys (and vendor defined) */
+	if (pkcs11_attr_class_is_key(class))
+		return PKCS11_CKR_TEMPLATE_INCONSISTENT;
+
+	rc = init_attributes_head(&obj2);
+	if (rc)
+		return rc;
+
+	/* Build a new serial object while sanitizing the attributes list */
+	rc = sanitize_client_object(&obj2, cur + sizeof(*cli_ref),
+				    cli_ref->size);
+	if (rc)
+		return rc;
+
+	return add_attribute(dst, cli_ref->id, obj2,
+			     sizeof(struct obj_attrs) + obj2->attrs_size);
+}
+
+enum pkcs11_rc sanitize_client_object(struct obj_attrs **dst, void *src,
+				      size_t size)
+{
+	struct pkcs11_object_head head = { };
+	enum pkcs11_rc rc = PKCS11_CKR_OK;
+	char *cur = NULL;
+	char *end = NULL;
+	size_t next = 0;
+
+	TEE_MemFill(&head, 0, sizeof(head));
+
+	if (size < sizeof(struct pkcs11_object_head))
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	TEE_MemMove(&head, src, sizeof(struct pkcs11_object_head));
+
+	if (size < (sizeof(struct pkcs11_object_head) + head.attrs_size))
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	rc = init_attributes_head(dst);
+	if (rc)
+		return rc;
+
+	rc = sanitize_class_and_type(dst, src);
+	if (rc)
+		goto bail;
+
+	rc = sanitize_boolprops(dst, src);
+	if (rc)
+		goto bail;
+
+	cur = (char *)src + sizeof(struct pkcs11_object_head);
+	end = cur + head.attrs_size;
+
+	for (; cur < end; cur += next) {
+		struct pkcs11_attribute_head cli_ref;
+
+		TEE_MemMove(&cli_ref, cur, sizeof(cli_ref));
+		next = sizeof(cli_ref) + cli_ref.size;
+
+		if (cli_ref.id == PKCS11_CKA_CLASS ||
+		    pkcs11_attr_is_type(cli_ref.id) ||
+		    pkcs11_attr_is_boolean(cli_ref.id))
+			continue;
+
+		rc = sanitize_indirect_attr(dst, &cli_ref, cur);
+		if (rc == PKCS11_CKR_OK)
+			continue;
+		if (rc != PKCS11_RV_NOT_FOUND)
+			goto bail;
+
+		if (!valid_pkcs11_attribute_id(cli_ref.id, cli_ref.size)) {
+			EMSG("Invalid attribute id %#"PRIx32, cli_ref.id);
+			rc = PKCS11_CKR_TEMPLATE_INCONSISTENT;
+			goto bail;
+		}
+
+		rc = add_attribute(dst, cli_ref.id, cur + sizeof(cli_ref),
+				   cli_ref.size);
+		if (rc)
+			goto bail;
+	}
+
+	/* sanity */
+	if (cur != end) {
+		EMSG("Unexpected alignment issue");
+		rc = PKCS11_CKR_FUNCTION_FAILED;
+		goto bail;
+	}
+
+bail:
+	return rc;
+}
+
+/*
+ * Debug: dump object attribute array to output trace
+ */
```

Comment:
Skipped review of the debugging helpers. Maybe wrap them in an `#ifdef DEBUG` so we don't accidentally provide code in the production binary that shouldn't be there. (This way we catch a forgotten ifdef-wrapping of a call of these functions.)